concrete-core-fixture/src/fixture/mod.rs

```diff
@@ -0,0 +1,179 @@
+//! A module containing a generic fixture trait for `concrete-core` operators.
+//! it
+//!
+//! The central abstraction of the library is the [`Fixture`] trait. This trait defines the logic
+//! used to sample / test / benchmark the implementors of any of the `*Engine` traits defined in
+//! `concrete-core`. This logic is always the same and, depending on the use of the fixture, boils
+//! down to:
+//!
+//! + Iterating over a set of parameters multiple times
+//! + Generating random raw inputs (where _raw_ ~ unsigned integers derivatives)
+//! + Preparing the operator pre-execution context using the raw inputs
+//! + Executing the operator in this context
+//! + Collecting the post-execution context, cleaning the entities, and extracting raw outputs
+//! + Predicting the statistical properties of the raw output sample using the raw inputs
+//! + Check that the output sample properties match the predictions
+//!
+//! For any given `*Engine` trait, a matching `*Fixture` type is defined, which _generically_
+//! implements [`Fixture`]. Here _generically_ means that the implementation of the [`Fixture`]
+//! trait do not fix the `Engine` and `Precision` type parameters, but rather restrict them to a
+//! family of types. In particular, the `Engine` generic type parameter must implement the `*Engine`
+//! trait in question. This `*Fixture` type can then be used to sample / test / benchmark any
+//! implementor of the `*Engine` trait.
+//!
+//! In particular, once the [`Fixture`] mandatory methods and types are defined, the user can
+//! benefit from the default methods [`Harness::sample`], [`Harness::test`] or [`Harness::stress`].
+use crate::generation::{IntegerPrecision, Maker};
+use crate::{Repetitions, SampleSize};
+use concrete_core::prelude::AbstractEngine;
+use std::ops::BitAnd;
+
+/// A trait for types implementing a fixture for a particular engine traits.
+///
+/// To understand how the differents pieces fit, see how the default methods `sample`, `test`,
+/// `stress` and `stress_all` use the associated types and methods.
+pub trait Fixture<Precision: IntegerPrecision, Engine: AbstractEngine, RelatedEntities> {
+    /// A type containing the parameters needed to generate input data.
+    type Parameters;
+
+    /// A type containing the raw inputs used to generate the execution context.
+    type RawInputs;
+
+    /// A type containing all the objects which must exist for the engine to be executed.
+    type PreExecutionContext;
+
+    /// A type containing prototypical values (such as secret keys) passed around the engine
+    /// execution.
+    type Bypass;
```

Comment:
calling it `Bypass` feels a bit strange, this type seems to be forwarded to `prepare_context` so I'd suggest `ForwardedContext` but I'm really not set.

Reply:
Actually the `Bypass` type contains prototypes of secret keys which needs to be forwarded to the `process_context` method to decrypt the output of the engine execution, but will _bypass_ the `execute_engine` method.

I would prefer not to use the `Context` word for this one, because it is not part of the execution context of the _engine_ (not forwarded to `execute_engine`). Something in the spirit of `Keys` or `SecretKeys` or `KeysProto` or `SecretKeysPrototypes` may be better suited. 

Waiting for your inputs on this one :wink:

Reply:
@aPere3 they're not key prototypes are they? They're `SynthesizedKeys` rather? Or do I get it wrong?

Reply:
No no, they are prototypes. Btw, if the `Prototypes` and `Synthesizing`  words do not work for you, I am completely open to something else .

Reply:
After today's discussion and your presentation I understand better :smile: Ok so `SecretKeyPrototypes` sounds good to me.

Reply:
Ok done !